libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,730 @@
+        .note("skipping invalid events")
+        .primary(args_.x)
+        .primary(args_.res->source)
+        .emit(dh);
+      return false;
+    }
+    return true;
+  }
+
+  auto validate_y(const data& d, tenzir::location loc,
+                  diagnostic_handler& dh) const -> bool {
+    if (args_.ty == chart_type::line and is<caf::none_t>(d)) {
+      return true;
+    }
+    const auto ty = type::infer(d);
+    TENZIR_ASSERT(ty);
+    if (not ty->kind()
+              .is_any<int64_type, uint64_type, double_type, duration_type>()) {
+      diagnostic::warning("Y-axis cannot have type `{}`", ty->kind())
+        .primary(std::move(loc))
+        .emit(dh);
+      return false;
+    }
+    return true;
+  }
+
+  // Modified from `floor()`
+  auto floor(const series& xs) const -> series {
+    return match(
+      *xs.array,
+      [&](const arrow::DurationArray& array) -> series {
+        auto b
+          = duration_type::make_arrow_builder(arrow::default_memory_pool());
+        check(b->Reserve(array.length()));
+        for (auto i = int64_t{0}; i < array.length(); i++) {
+          if (array.IsNull(i)) {
+            check(b->AppendNull());
+            continue;
+          }
+          const auto val = array.Value(i);
+          const auto count = std::abs(args_.res->inner.count());
+          const auto rem = std::abs(val % count);
+          if (rem == 0) {
+            check(b->Append(val));
+            continue;
+          }
+          const auto floor = val >= 0 ? -rem : rem - count;
+          check(b->Append(val + floor));
+        }
+        return {duration_type{}, finish(*b)};
+      },
+      [&](const arrow::TimestampArray& array) -> series {
+        auto opts = make_round_temporal_options(args_.res->inner);
+        return {time_type{},
+                check(arrow::compute::FloorTemporal(array, std::move(opts)))
+                  .array_as<arrow::TimestampArray>()};
+      },
+      [&](const auto&) -> series {
+        TENZIR_UNREACHABLE();
+      });
+  }
+
+  friend auto inspect(auto& f, chart_operator2& x) -> bool {
+    return f.object(x)
+      .pretty_name("chart_operator2")
+      .fields(f.field("args_", x.args_));
+  }
+
+private:
+  chart_args args_;
+};
+
+template <chart_type Ty>
+class chart_plugin : public virtual operator_plugin2<chart_operator2> {
+  auto name() const -> std::string override {
+    return fmt::format("chart_{}", to_string(Ty));
+  }
+
+  auto make(invocation inv, session ctx) const
+    -> failure_or<operator_ptr> override {
+    auto args = chart_args{};
+    args.ty = Ty;
+    auto y = ast::expression{};
+    auto xmin = std::optional<ast::expression>{};
+    auto xmax = std::optional<ast::expression>{};
+    auto p = argument_parser2::operator_(name());
+    p.named(Ty == chart_type::pie ? "names" : "x", args.x, "any");
+    p.named(Ty == chart_type::pie ? "values" : "y", y, "any");
+    p.named("x_min", xmin, "constant");
+    p.named("x_max", xmax, "constant");
```

Comment:
This didn't respect `x_min` and `x_max` as expected:

```
// tql2
metrics "operator"
where source and pipeline_id == "ed37bfa8-2724-4180-a833-39282bc1bbb1"
chart_area x=timestamp, y=processing_duration, x_min=now(), x_max=now()+1min
```

Also not respected with `chart_line`, but there the chart is completely gone.